q2_dada2/_denoise.py

```diff
@@ -6,28 +6,93 @@
 # The full license is in the file LICENSE, distributed with this software.
 # ----------------------------------------------------------------------------
 
-import os.path
+import os
 import tempfile
 import hashlib
 
 import biom
 import skbio
 from q2_types.feature_data import DNAIterator
 from q2_types.per_sample_sequences import (
-    SingleLanePerSampleSingleEndFastqDirFmt)
+    FastqGzFormat, SingleLanePerSampleSingleEndFastqDirFmt,
+    SingleLanePerSamplePairedEndFastqDirFmt)
 
 from q2_dada2._plot import run_commands
 
 
-def denoise(demultiplexed_seqs: SingleLanePerSampleSingleEndFastqDirFmt,
-            trunc_len: int, trim_left: int, max_ee: int=2, truncq: int=2,
-            hashed_feature_ids: bool=True) -> (biom.Table, DNAIterator):
+def _check_featureless_table(fp):
+    with open(fp) as fh:
+        # There is a comment line and a header before the feature data
+        for line_count, _ in zip(range(3), fh):
+            pass
+    if line_count < 2:
+        raise ValueError("No features remain after denoising. "
+                         "Try adjusting your arguments.")
+
+
+def denoise_single(demultiplexed_seqs: SingleLanePerSampleSingleEndFastqDirFmt,
+                   trunc_len: int, trim_left: int, max_ee: float=2.5,
+                   trunc_q: int=2, n_threads: int=1,
+                   n_reads_learn: int=1000000, hashed_feature_ids: bool=True
```

Comment:
The script will use the first X samples, where X is at least 1, and the number of reads in X is at least greater than `n_reads_learn`, to learn the error rates.

So `n_reads_learn=0` will result in just 1 sample, the first sample, being used to estimate the error rates.

*`n_reads_learn` is really a minimum number of reads, not an exact number*